### PR TITLE
Fixed gradient tracking

### PR DESCRIPTION
**Pull request by pdanilov**
## Description

Fixed storing gradients in OptimizerCallback

## Type of Change

- [ ] Examples / docs / tutorials / contributors update
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves an existing feature)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] I have read the [Code of Conduct](https://github.com/catalyst-team/catalyst/blob/master/CODE_OF_CONDUCT.md) document.
- [x] I have read the [Contributing](https://github.com/catalyst-team/catalyst/blob/master/CONTRIBUTING.md) guide.
- [x] I have checked the code-style using `make check-codestyle`.
- [x] I have written tests for all new methods and classes that I created.
- [x] I have written the docstring in Google format for all the methods and classes that I used.
- [x] I have checked the docs using `make check-docs`.
- [x] I have read I need to click 'Login as guest' to see Teamcity build logs.

### CODE REVIEW

**Comment by Scitator** [comment]
Dear @pdanilov , thanks for your contribution,
could you please write any test to check, that callback works correctly? and also show how it should work :)

**Comment by pdanilov** [comment]
What is the location for tests of callbacks?
I can only locate catalyst/utils/tests/ in which there are no tests for callbacks

**Comment by Scitator** [comment]
you can create ones into `catalyst/core/callbacks/tests/*.py`
there should be more someday :)

**Comment by pdanilov** [comment]
I'm in trouble with tests, because I don't understand how to forward some dummy experiment class (for example, from cifar-simple) into SupervisedRunner, when we deal not with configs, but with the core code of framework.

**Comment by Scitator** [comment]
@pdanilov you can use Experiment test as an example (`https://github.com/catalyst-team/catalyst/blob/master/catalyst/dl/experiment/tests/test_supervised.py`)
and just do something like
```
runner = SupervisedRunner()
runner.experiment = SupervisedExperiment()
runner._prepare_for_stage("train")

batch = ...
runner._run_batch(batch)

// check runner.state

```

**Comment by pdanilov** [comment]
Yeah, that's exactly what I am looking for, but Catalyst doesn't provide mechanics like that: you cannot set experiment field into your runner. Your runner can only create local variable experiment as an instance of `_default_experiment`, and then run this experiment inside train method.

https://github.com/catalyst-team/catalyst/blob/master/catalyst/dl/runner/supervised.py#L199-L218

**Comment by pdanilov** [comment]
@Scitator any ideas?

**Comment by Scitator** [comment]
Sorry, could not understand you issue. You can use Experiment or State to store whatever you need.
Moreover, the example code is fully working one.

PS. btw, could you please merge with current master?

**Comment by Scitator** [comment]
@pdanilov could you please merge with master?
we have a bunch of updates :)

**Comment by pdanilov** [comment]
ok

**Comment by Scitator** [comment]
@pdanilov could you please fix codestyle issues?

**Comment by pdanilov** [comment]
Ready for merge

**Comment by pdanilov** [comment]
@Scitator merge?

File: catalyst/core/callbacks/optimizer.py [hunk 1]
```diff
@@ -172,14 +167,88 @@ def on_batch_end(self, state: State) -> None:
...
+        need_gradient_step = (
+            self._accumulation_counter % self.accumulation_steps == 0
+        )
+
+        if need_gradient_step:
+            self.save_grad_norm(
+                model=state.model,
+                prefix=self.grad_norm_prefix,
+                norm_type=self.norm_type,
+                output=state.batch_metrics,
```
Comment (Scitator) [inline catalyst/core/callbacks/optimizer.py]: please don't use python inplace return, let's use something like
```
output = self.save_grad_norm(
                model=state.model,
                prefix=self.grad_norm_prefix,
                norm_type=self.norm_type)
# update state.batch_metrics with output
```

File: catalyst/core/callbacks/tests/test_optimizer_callback.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+
+    runner = SupervisedRunner()
+    runner.train(
+        model=model,
+        criterion=criterion,
+        optimizer=optimizer,
+        loaders=loaders,
+        logdir=logdir,
+        callbacks=callbacks,
+    )
```
Comment (Scitator) [inline catalyst/core/callbacks/tests/test_optimizer_callback.py]: I suppose you can use `check=True` during tests

**Review by Scitator — CHANGES_REQUESTED** [review]

**Comment by Scitator** [comment]
btw, could you please also move Callaback to `https://github.com/catalyst-team/catalyst/tree/master/catalyst/contrib/dl/callbacks/optimizer.py`? I think, we need to check it for some time, before moving directly to Core.

**Comment by Scitator** [comment]
@pdanilov Thank you for you contribution!

**Pull request merged**